README
		KXL, a visual,sound library
		Copyright (C) 2001, Katsuyoshi Sato
			fc3srx7@mwnet.or.jp

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
675 Mass Ave, Cambridge, MA 02139, USA.
--------------------------------------------------------------------------
#Install

  tar xzvf KXL-xxxx(xxxx is version)
  cd KXL-xxxx
  ./configure
  make
  make install
--------------------------------------------------------------------------
#Requirements
pulseaudio
libpulse-devel
libX11-devel
--------------------------------------------------------------------------

Note: The sources use the euc-jp encoding. To use this encoding in Vim,
edit a file and then run this command to switch the encoding:
:edit ++encoding=euc-jp


ChangeLog
Unreleased Version 1.2.5 [9/9/2023]
	- KXL_ReadBitmapHeader no longer loops indefinitely on corrupted BMPs.
	- KXL_LoadBitmap and KXL_ReadBitmapHeader will fail instead of exiting your program.
	- Documented KXL_ReadBitmapHeader.
	- Added AFL++ sample.c fuzz target in the test/ directory.
	- More fuzzing friendly because there are now fewer exit calls.
-------------------------------------------------------------------------------
Release Version 1.2.4 [30/9/2021]
	- Fix incompatible pointer type warning.
-------------------------------------------------------------------------------
Release Version 1.2.3-1 [29/9/2021]
	- Fix some pointer sign warnings
-------------------------------------------------------------------------------
Release version 1.2.2-1 [6/3/2020]
	- Fix Valgrind warning "Invalid read of size 8" in KXL_PlaySound
	- Zero bitmap data to silence Valgrind warning
-------------------------------------------------------------------------------
Release Version 1.2.1-1 [5/3/2020]
	- Fix Valgrind warning write(buf) points to uninitialised byte(s)
-------------------------------------------------------------------------------
Release Version 1.2.0-2 [4/3/2020]
	- Fix rpmbuild failing to find KXL library in lib64.
-------------------------------------------------------------------------------
Release Version 1.2.0-1 [4/3/2020]
	- Use PulseAudio by default for sound output.
-------------------------------------------------------------------------------
Release Version 1.1.6-1 [22/12/2019]
	- Change obsolete Copyright tag to License tag
	- Fix typo in format string.
	- Fix commented out KXL_KEY_Bracketright
	- Fix "implicit declaration of built-in function" warnings
	- Fix a conditional jump or move error from Valgrind
	- Fix approx 1MB memory leak in KXL_Font()
-------------------------------------------------------------------------------
Release Version 1.1.5 [19/1/2002]
	- KXL_UpDateRect function was added.
	- KXL_UpDateImm function was added.
	- KXL_ClearFrameRect function was added.
	- KXL_ClearFrameImm function was added.
	- KXL_CopyImageRect function was added.
	- KXL_CopyImageImm function was added.
	- KXL_CopyStretchImageRect function was added.
	- KXL_CopyStretchImageImm function was added.
	- KXL_PutImageRect function was added.
	- KXL_PutImageImm function was added.
	- KXL_PutStretchImageRect function was added.
	- KXL_PutStretchImageImm function was added.
	- Improvement in the speed of various functions.
-------------------------------------------------------------------------------
Release Version 1.1.4 [8/12/2001]
	- The bug in which the expansion image has collapsed was corrected
	with the KXL_StrechImage function.
-------------------------------------------------------------------------------
Release Version 1.1.3 [10/09/2001]
	- When sound was sounded with single shot, operation corrected the bug which becomes blunt.
	- Painting-out processing was added to the KXL_DrawPolygon function.
-------------------------------------------------------------------------------
Release Version 1.1.2 [29/08/2001]
	- spec file added.
-------------------------------------------------------------------------------
Release Version 1.1.1 [15/08/2001]
	- The bug of KXL.m4 was corrected.
-------------------------------------------------------------------------------
Release Version 1.1.0 [13/08/2001]
	- 4bpp bitmap iamge support.
	- change name Timmerxxx to Timerxxx.
	- speed tuning.
-------------------------------------------------------------------------------
Release Version 1.0.0 [12/08/2001]
	- Major Upgrade.
	- KXL_SetDrawColor,KXL_DrawPoint,KXL_DrawLine,KXL_DrawRectangle
	KXL_DrawPolygon added.
	- libxpm was removed from the library.
	- An old game does not operate in the version from this time.
-------------------------------------------------------------------------------
Release Version 0.5.8 [08/08/2001]
        - The mouse event was supported.
	- Bitmap file supported.
-------------------------------------------------------------------------------
Release Version 0.5.7 [26/07/2001]
	- The header file was corrected.
-------------------------------------------------------------------------------
Release Version 0.5.6 [17/05/2001]
	- The function which creates an expansion reduction image was added.
-------------------------------------------------------------------------------
Release Version 0.5.5 [07/04/2001]
	- It was made to correspond to configure script.
-------------------------------------------------------------------------------
Release Version 0.5.4 [04/04/2001]
	- Makefile was corrected.
-------------------------------------------------------------------------------
Release Version 0.5.3 [28/03/2001]
	- It is at KXL_PlaySound at the KXL_SOUND_STOP and
	  KXL_SOUND_STOP_ALL specification time.
	- The fault sound was sounding once was corrected.
-------------------------------------------------------------------------------
Release Version 0.5.2 [24/03/2001]
	- Script which acquires the path of a header & library in
	  KXL It corrected.
-------------------------------------------------------------------------------
Release Version 0.5.1 [11/03/2001]
	- It is the fault to which clipmask was not correctly carried out
	  with a KXL_PutRectImage function.
	- It corrected.
-------------------------------------------------------------------------------
Release Version 0.5 [05/03/2001]
	- Fault was corrected.
-------------------------------------------------------------------------------
Release Version 0.4 [28/02/2001]
	- Change accompanied by renewal of a library was carried out.
	- Fault was corrected.
-------------------------------------------------------------------------------
Release Version 0.3 [25/02/2001]
	- first (pre)release version
	- It is a multimedia library aiming at easy game development.
	- From now on, all the things raised here will use this library.
	- I want to explain a library gradually.

test/sample.c
// sample.c
#include <KXL.h>

int main(int argc, char** argv)
{
  Bool flag = True;

  // Create a window by 100x100.
  // event receives only the bottom of a re-drawing event and key presss.
  // If return key is pushed, it will end.
  KXL_CreateWindow(100, 100, "sample",
                   KXL_EVENT_EXPOSURE_MASK |
                   KXL_EVENT_KEY_PRESS_MASK);
  KXL_Image *theImage = KXL_LoadBitmap(argv[1], 0);

  while(flag != True ){
    // event loop
    while(KXL_CheckEvents()) {
      // which acquires a event
      switch(KXL_GetEvents()) {
      case KXL_EVENT_EXPOSE: // event of re-drawing
        // A frame is cleared
        KXL_ClearFrameImm(0, 0, 100, 100);
        // "sample" is drawn
        KXL_PutImage(theImage, 0, 0);
        // A frame is made to reflect in a window
        KXL_UpDateImm(0, 0, 100, 100);
        break;
      case KXL_EVENT_KEY_PRESS: // event of key press
        // If a return key was pushed, it will end
        if (KXL_GetKey() == KXL_KEY_Return)
          flag = True;
        break ;
      }
    }
  }
  KXL_DeleteImage(theImage);
  // A window is closed
  KXL_DeleteWindow();
  return 0;
}

tests/test_001.c
#include <stdlib.h>
#include <check.h>
#include <KXL.h>

extern KXL_Window *KXL_Root;
extern Bool KXL_SoundOk;

START_TEST (test_CreateWindow)
{
    KXL_CreateWindow(100, 100, "CreateWindow test",
                   KXL_EVENT_EXPOSURE_MASK |
                   KXL_EVENT_KEY_PRESS_MASK);

    ck_assert( KXL_Root->Display != NULL );
    KXL_DeleteWindow();

    KXL_DisplayName("abcdefghijklmnopqrstuvwxyz");
    KXL_CreateWindow(100, 100, "CreateWindow test",
                   KXL_EVENT_EXPOSURE_MASK |
                   KXL_EVENT_KEY_PRESS_MASK);
    ck_assert(!KXL_Root->Display);
}
END_TEST

START_TEST(test_LoadBitmap)
{
    KXL_CreateWindow(100, 100, "CreateWindow test",
                   KXL_EVENT_EXPOSURE_MASK |
                   KXL_EVENT_KEY_PRESS_MASK);

    KXL_Image *bmp = KXL_LoadBitmap("../../geki2/bmp/boss1.bmp", 0);
    ck_assert(bmp != NULL);
    KXL_DeleteImage(bmp);

    bmp = KXL_LoadBitmap("", 0);
    ck_assert(!bmp);
    KXL_DeleteWindow();
}
END_TEST

START_TEST(test_InitSound)
{
    char *snames[] = {"bgm1", ""};
    KXL_InitSound("../../geki2/wav", snames);
    ck_assert(KXL_SoundOk);
    KXL_EndSound();
    char *notexist[] = {"non-existent", ""};
    KXL_InitSound("/dev/random/foo", notexist);
    ck_assert(!KXL_SoundOk);
    KXL_EndSound();
}
END_TEST

Suite * KXL_suite(void)
{
    Suite *s;
    TCase *KXL_core;

    s = suite_create("KXL tests");
    KXL_core = tcase_create("Core tests");

    tcase_add_test(KXL_core, test_CreateWindow);
    tcase_add_test(KXL_core, test_LoadBitmap);
    tcase_add_test(KXL_core, test_InitSound);

    suite_add_tcase(s, KXL_core);

    return s;
}

int main(void)
{
    int number_failed;
    Suite *s;
    SRunner *sr;

    s = KXL_suite();
    sr = srunner_create(s);

    srunner_run_all(sr, CK_NORMAL);
    number_failed = srunner_ntests_failed(sr);
    srunner_free(sr);
    return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
}

test/libFuzzer-target.c
#ifdef __cplusplus
extern "C" {
#endif
#include <KXL.h>
#ifdef __cplusplus
}
#endif
#include <stdlib.h>
#include <stdint.h>
#include <unistd.h>
#include <sys/types.h>
#include <assert.h>

static FILE* temp = NULL;
static char* tempname = NULL;


void fuzzCleanup(void);
char Template[] = "test_bmp_XXXXXX";

#ifdef __cplusplus
extern "C"
#endif
int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
    /* If this is the first call of this function, create temp file. */
    if (!temp) {
        tempname = mktemp(Template);
        if (!tempname)
            perror("tempname is NULL");
        temp = fopen(tempname, "w+");
        if (!temp) {
            perror("temp is NULL");
            printf("tempname=%s\n", tempname);
        }
        KXL_CreateWindow(800,600,"fuzz test",KXL_EVENT_EXPOSURE_MASK |
                   KXL_EVENT_KEY_PRESS_MASK);
        atexit(fuzzCleanup);
    }

    assert(temp);

    if (Size == 0)
        return 0;

    /* First byte is the blend value */
    uint8_t blend = Data[0];

    /* Write to the file */
    truncate(tempname, Size);
    rewind(temp);
    fwrite(Data+1, Size-1, 1, temp);
    fflush(temp);

    /* Call function under test */
    KXL_Image* img = KXL_LoadBitmap(tempname, blend);

    /* Cleanup so we don't leak memory */
    if (img)
        KXL_DeleteImage(img);
    return 0;
}
void fuzzCleanup(void) {
    KXL_DeleteWindow();
    fclose(temp);
    unlink(tempname);
}

src/KXLmisc.c
#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>
#include <signal.h>
#include <sys/time.h>
#include "KXL.h"

Bool KXL_TimerFlag;

// 360度データ
static Sint16 sin360[] = {
    0,
    4,    8,   13,   17,   22,   26,   31,   35,   40,   44,
   48,   53,   57,   61,   66,   70,   74,   79,   83,   87,
   91,   95,  100,  104,  108,  112,  116,  120,  124,  127,
  131,  135,  139,  143,  146,  150,  154,  157,  161,  164,
  167,  171,  174,  177,  181,  184,  187,  190,  193,  196,
  198,  201,  204,  207,  209,  212,  214,  217,  219,  221,
  223,  226,  228,  230,  232,  233,  235,  237,  238,  240,
  242,  243,  244,  246,  247,  248,  249,  250,  251,  252,
  252,  253,  254,  254,  255,  255,  255,  255,  255,  256,
  255,  255,  255,  255,  255,  254,  254,  253,  252,  252,
  251,  250,  249,  248,  247,  246,  244,  243,  242,  240,
  238,  237,  235,  233,  232,  230,  228,  226,  223,  221,
  219,  217,  214,  212,  209,  207,  204,  201,  198,  196,
  193,  190,  187,  184,  181,  177,  174,  171,  167,  164,
  161,  157,  154,  150,  146,  143,  139,  135,  131,  127,
  124,  120,  116,  112,  108,  104,  100,   95,   91,   87,
   83,   79,   74,   70,   66,   61,   57,   53,   48,   44,
   40,   35,   31,   26,   22,   17,   13,    8,    4,    0,
   -4,   -8,  -13,  -17,  -22,  -26,  -31,  -35,  -40,  -44,
  -48,  -53,  -57,  -61,  -66,  -70,  -74,  -79,  -83,  -87,
  -91,  -95, -100, -104, -108, -112, -116, -120, -124, -127,
 -131, -135, -139, -143, -146, -150, -154, -157, -161, -164,
 -167, -171, -174, -177, -181, -184, -187, -190, -193, -196,
 -198, -201, -204, -207, -209, -212, -214, -217, -219, -221,
 -223, -226, -228, -230, -232, -233, -235, -237, -238, -240,
 -242, -243, -244, -246, -247, -248, -249, -250, -251, -252,
 -252, -253, -254, -254, -255, -255, -255, -255, -255, -256,
 -255, -255, -255, -255, -255, -254, -254, -253, -252, -252,
 -251, -250, -249, -248, -247, -246, -244, -243, -242, -240,
 -238, -237, -235, -233, -232, -230, -228, -226, -223, -221,
 -219, -217, -214, -212, -209, -207, -204, -201, -198, -196,
 -193, -190, -187, -184, -181, -177, -174, -171, -167, -164,
 -161, -157, -154, -150, -146, -143, -139, -135, -131, -128,
 -124, -120, -116, -112, -108, -104, -100,  -95,  -91,  -87,
  -83,  -79,  -74,  -70,  -66,  -61,  -57,  -53,  -48,  -44,
  -40,  -35,  -31,  -26,  -22,  -17,  -13,   -8,   -4,    0,
};

//==============================================================
//  タイマーフラグ取得
//==============================================================
Bool KXL_GetTimer(void)
{
  return KXL_TimerFlag;
}

//==============================================================
//  タイマーフラグリセット
//==============================================================
void KXL_ResetTimer(void)
{
  KXL_TimerFlag = False;
}

//==============================================================
//  タイマーコールバック
//==============================================================
void KXL_TimerCallBack(int dummy)
{
  KXL_TimerFlag = True;
}

//==============================================================
//  タイマー設定
//  引き数：フレーム数
//==============================================================
void KXL_Timer(Uint16 time)
{
  struct itimerval val={
    {0, 1000000 / time},
    {0, 1000000 / time},
  };
  signal(SIGALRM, KXL_TimerCallBack);
  setitimer(ITIMER_REAL, &val, NULL);
  KXL_TimerFlag = False;
}

//==============================================================
//  メモリ確保
//  引き数：メモリサイズ
//  戻り値：確保したメモリのポインタ
//==============================================================
void *KXL_Malloc(Uint32 size)
{
  void *new;

  new = malloc(size);
  if (new == NULL) {
    fprintf(stderr, "KXL error message\nOut Of memory!!\n");
    exit(1);
  }
  return new;
}

//==============================================================
//  メモリ再確保
//  引き数：メモリのポインタ
//        ：メモリサイズ
//  戻り値：再確保したメモリのポインタ
//==============================================================
void *KXL_Realloc(void *src, Uint32 size)
{
  void *new;

  new = realloc(src, size);
  if (new == NULL) {
    fprintf(stderr, "KXL error message\nOut Of memory!!\n");
    free(src);
    exit(1);
  }
  return new;
}

//==============================================================
//  メモリ解放
//  引き数：メモリのポインタ
//==============================================================
void KXL_Free(void *src)
{
  free(src);
}

//==============================================================
//  方角取得
//  引き数：自分の矩形
//        ：相手の矩形
//==============================================================
Uint16 KXL_GetDirection(KXL_Rect src, KXL_Rect target) {
  Uint16 k, x, y;
  Uint16 mx, my, yx, yy;

  mx = src.Left + (src.Width >> 1);
  my = src.Top + (src.Height >> 1);
  yx = target.Left + (target.Width >> 1);
  yy = target.Top + (target.Height >> 1);

  x = abs(yx - mx);
  y = abs(yy - my);
  if (yx == mx)
    k = (yy > my) ? 0 : 180;
  else if (yy == my)
    k = (yx > mx) ? 90 : 270;
  else if (yx > mx)
    if (yy > my)
      k = 90 * x / (x + y);
    else
      k = 180 - (90 * x / (x + y));
  else
    if (yy > my)
      k = 360 - (90 * x / (x + y));
    else
      k = (90 * x / (x + y)) + 180;
  return k;
}

//==============================================================
//  角度による加算値設定
//  引き数：角度
//        ：水平加算値のポインタ
//        ：垂直加算値のポインタ
//==============================================================
void KXL_GetDirectionAdd(Sint16 dir, Sint16 *x, Sint16 *y) {
  Sint16 dir2 = dir + 90;

  while (dir < 0) dir += 360;
  while (dir > 360) dir -= 360;
  *x = sin360[dir];

  while (dir2 < 0) dir2 += 360;
  while (dir2 > 360) dir2 -= 360;
  *y = sin360[dir2];
}

//==============================================================
//  １６リトルビットエンディアン読み込み
//  引き数：ファイルポインタ
//  戻り値：１６ビット値
//==============================================================
Uint16 KXL_ReadU16(FILE *fp)
{
  Uint8 c[2];

  fread(c, 1, 2, fp);
  return (Uint16)(c[0] +  c[1] * 0x100);
}

//==============================================================
//  ３２ビットリトルエンディアン読み込み
//  引き数：ファイルポインタ
//  戻り値：３２ビット値
//==============================================================
Uint32 KXL_ReadU32(FILE *fp)
{
  Uint8 c[4];

  fread(c, 1, 4, fp);
  return (Uint32)(c[0] + c[1] * 0x100L + c[2] * 0x10000L + c[3] * 0x1000000L);
}



src/KXL.h
#ifndef _KXLIB_H_
#define _KXLIB_H_

#include <stdio.h>
#include <stdint.h>
#include <limits.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>

//================================================================
// Global structures and variables
//================================================================
// Event mask
#define KXL_EVENT_NO_MASK               0L
#define KXL_EVENT_KEY_PRESS_MASK        (1L<<0)
#define KXL_EVENT_KEY_RELEASE_MASK      (1L<<1)
#define KXL_EVENT_BUTTON_PRESS_MASK     (1L<<2)
#define KXL_EVENT_BUTTON_RELEASE_MASK   (1L<<3)
#define KXL_EVENT_BUTTON_MOTION_MASK    (1L<<13)
#define KXL_EVENT_EXPOSURE_MASK	        (1L<<15) 
// Event
#define KXL_EVENT_KEY_PRESS       2
#define KXL_EVENT_KEY_RELEASE     3
#define KXL_EVENT_BUTTON_PRESS    4
#define KXL_EVENT_BUTTON_RELEASE  5
#define KXL_EVENT_BUTTON_MOTION   6
#define KXL_EVENT_EXPOSE         12
// Key code
#define KXL_KEY_BackSpace       0xff08
#define KXL_KEY_Tab             0xff09
#define KXL_KEY_Return          0xff0d
#define KXL_KEY_Escape          0xff1b
#define KXL_KEY_Delete          0xffff
#define KXL_KEY_Home            0xff50
#define KXL_KEY_Left            0xff51
#define KXL_KEY_Up              0xff52
#define KXL_KEY_Right           0xff53
#define KXL_KEY_Down            0xff54
#define KXL_KEY_F1              0xffbe
#define KXL_KEY_F2              0xffbf
#define KXL_KEY_F3              0xffc0
#define KXL_KEY_F4              0xffc1
#define KXL_KEY_F5              0xffc2
#define KXL_KEY_F6              0xffc3
#define KXL_KEY_F7              0xffc4
#define KXL_KEY_F8              0xffc5
#define KXL_KEY_F9              0xffc6
#define KXL_KEY_F10             0xffc7
#define KXL_KEY_F11             0xffc8
#define KXL_KEY_F12             0xffc9
#define KXL_KEY_Shift_L	        0xffe1
#define KXL_KEY_Shift_R	        0xffe2
#define KXL_KEY_Control_L       0xffe3
#define KXL_KEY_Control_R       0xffe4
#define KXL_KEY_Space           0x020 // 
#define KXL_KEY_Exclam          0x021 // !
#define KXL_KEY_Quotedbl        0x022 // "
#define KXL_KEY_Numbersign      0x023 // #
#define KXL_KEY_Dollar          0x024 // $
#define KXL_KEY_Percent         0x025 // %
#define KXL_KEY_Ampersand       0x026 // &
#define KXL_KEY_Apostrophe      0x027 //
#define KXL_KEY_Quoteright      0x027 // 
#define KXL_KEY_Parenleft       0x028 // (
#define KXL_KEY_Parenright      0x029 // )
#define KXL_KEY_Asterisk        0x02a // *
#define KXL_KEY_Plus            0x02b // +
#define KXL_KEY_Comma           0x02c // ,
#define KXL_KEY_Minus           0x02d // -
#define KXL_KEY_Period          0x02e // .
#define KXL_KEY_Slash           0x02f // /
#define KXL_KEY_0               0x030
#define KXL_KEY_1               0x031
#define KXL_KEY_2               0x032
#define KXL_KEY_3               0x033
#define KXL_KEY_4               0x034
#define KXL_KEY_5               0x035
#define KXL_KEY_6               0x036
#define KXL_KEY_7               0x037
#define KXL_KEY_8               0x038
#define KXL_KEY_9               0x039
#define KXL_KEY_Colon           0x03a // :
#define KXL_KEY_Semicolon       0x03b // ;
#define KXL_KEY_Less            0x03c // <
#define KXL_KEY_Equal           0x03d // =
#define KXL_KEY_Greater         0x03e // >
#define KXL_KEY_Question        0x03f // ?
#define KXL_KEY_At              0x040 // @
#define KXL_KEY_A               0x041
#define KXL_KEY_B               0x042
#define KXL_KEY_C               0x043
#define KXL_KEY_D               0x044
#define KXL_KEY_E               0x045
#define KXL_KEY_F               0x046
#define KXL_KEY_G               0x047
#define KXL_KEY_H               0x048
#define KXL_KEY_I               0x049
#define KXL_KEY_J               0x04a
#define KXL_KEY_K               0x04b
#define KXL_KEY_L               0x04c
#define KXL_KEY_M               0x04d
#define KXL_KEY_N               0x04e
#define KXL_KEY_O               0x04f
#define KXL_KEY_P               0x050
#define KXL_KEY_Q               0x051
#define KXL_KEY_R               0x052
#define KXL_KEY_S               0x053
#define KXL_KEY_T               0x054
#define KXL_KEY_U               0x055
#define KXL_KEY_V               0x056
#define KXL_KEY_W               0x057
#define KXL_KEY_X               0x058
#define KXL_KEY_Y               0x059
#define KXL_KEY_Z               0x05a
#define KXL_KEY_Bracketleft     0x05b // "
#define KXL_KEY_Backslash       0x05c // '\'
#define KXL_KEY_Bracketright    0x05d // #
#define KXL_KEY_Asciicircum     0x05e
#define KXL_KEY_Underscore      0x05f // _
#define KXL_KEY_Grave           0x060
#define KXL_KEY_Quoteleft       0x060
#define KXL_KEY_a               0x061
#define KXL_KEY_b               0x062
#define KXL_KEY_c               0x063
#define KXL_KEY_d               0x064
#define KXL_KEY_e               0x065
#define KXL_KEY_f               0x066
#define KXL_KEY_g               0x067
#define KXL_KEY_h               0x068
#define KXL_KEY_i               0x069
#define KXL_KEY_j               0x06a
#define KXL_KEY_k               0x06b
#define KXL_KEY_l               0x06c
#define KXL_KEY_m               0x06d
#define KXL_KEY_n               0x06e
#define KXL_KEY_o               0x06f
#define KXL_KEY_p               0x070
#define KXL_KEY_q               0x071
#define KXL_KEY_r               0x072
#define KXL_KEY_s               0x073
#define KXL_KEY_t               0x074
#define KXL_KEY_u               0x075
#define KXL_KEY_v               0x076
#define KXL_KEY_w               0x077
#define KXL_KEY_x               0x078
#define KXL_KEY_y               0x079
#define KXL_KEY_z               0x07a
#define KXL_KEY_Braceleft       0x07b
#define KXL_KEY_Bar             0x07c
#define KXL_KEY_Braceright      0x07d
#define KXL_KEY_Asciitilde      0x07e // ~
// Type of variable
typedef signed char    Sint8;
typedef unsigned char  Uint8;
typedef signed short   Sint16;
typedef unsigned short Uint16;
typedef int32_t Sint32;
typedef uint32_t Uint32;
// Image
typedef struct {
  Pixmap Buffer;         // Drawing image
  Pixmap Mask;           // Mask image
  GC     MaskGC;         // GC of mask image
  Uint16 Width;          // Drawing width
  Uint16 Height;         // Drawing height
} KXL_Image;
// Rectangle
typedef struct {
  Sint16 Left;           // Left
  Sint16 Top;            // Top
  Uint16 Width;          // Width
  Uint16 Height;         // Height
} KXL_Rect;
// Polygon
typedef XPoint KXL_Polygon;
// Sound control
typedef enum {
  KXL_SOUND_PLAY,        // Play sound
  KXL_SOUND_PLAY_LOOP,   // Loop play sound
  KXL_SOUND_STOP,        // Stop sound
  KXL_SOUND_STOP_ALL     // Stop all sound
} KXL_Command;

//================================================================
// Internal structures
//================================================================
// Frame
typedef struct {
  Pixmap    Buffer;      // Buffer
  GC        Gc;          // GC
  Uint16    Width;       // Width
  Uint16    Height;      // Height
} KXL_Frame;
// Window
typedef struct {
  Display     *Display;  // Pointer of display structure
  Window       Win;      // window
  Uint16       Scr;      // Screen number
  XEvent       Event;    // Event of X window
  Uint16       Depth;    // Bits per pixel of X window
  Colormap     Cmap;     // Color map
  Uint16       Width;    // Width
  Uint16       Height;   // Height
  KXL_Frame   *Frame;    // Pointer of frame structure
  GC           FontGC;   // GC of font
  XFontStruct *WinFont;  // Pointer of font structrue
} KXL_Window;
// RGBE palette
typedef struct {
  Uint8 b, g, r, e;      // Blue, Green, Red, ?
} KXL_RGBE;
// Bitmap header
typedef struct {
  Uint8     magic[2];    // Bitmap magic
  Uint32    file_size;   // File size
  Uint16    reserved1;   // Reserved
  Uint16    reserved2;   // Reserved
  Uint32    offset;      // Offset of image
  Uint32    hed_size;    // Header size
  Uint32    width;       // Image width
  Uint32    height;      // Image height
  Uint16    plane;       // Plane
  Uint16    depth;       // Bits per pixel
  Uint32    lzd;         // Compression
  Uint32    image_size;  // Image size
  Uint32    x_pixels;    // Horizon resolution
  Uint32    y_pixels;    // Vertical resolution
  Uint32    pals;        // Palette
  Uint32    pals2;       // Palette
  KXL_RGBE *rgb;         // Palette data
  Uint32    w;           // Bytes per line
  Uint8    *data;        // Image data
} KXL_BitmapHeader;

//================================================================
// Internal function prototypes
//================================================================
void        KXL_Up_Date(Uint16 src_l, Uint16 src_t, Uint16 src_w, Uint16 src_h);
void        KXL_Clear_Frame(Uint16 src_l, Uint16 src_t, Uint16 src_w, Uint16 src_h);
KXL_Image  *KXL_Copy_Image(KXL_Image *src, Uint16 src_l, Uint16 src_t, Uint16 src_w, Uint16 src_h);
KXL_Image  *KXL_Copy_StretchImage(KXL_Image *src, Uint16 src_l, Uint16 src_t, Uint16 src_w, Uint16 src_h, Uint16 width, Uint16 height);
void        KXL_Put_Image(KXL_Image *img, Uint16 src_l, Uint16 src_t, Uint16 src_w, Uint16 src_h, Sint16 x, Sint16 y);
void        KXL_Put_StretchImage(KXL_Image *img, Uint16 src_l, Uint16 src_t, Uint16 src_w, Uint16 src_h, Uint16 w, Uint16 h, Sint16 x, Sint16 y);
void        KXL_CreateBitmap8to16(Uint8 *from, XImage *to, KXL_RGBE *rgb, Uint8 blend);
void        KXL_CreateBitmap8to24(Uint8 *from, XImage *to, KXL_RGBE *rgb, Uint8 blend);
void        KXL_CreateBitmap8to1(Uint8 *from, XImage *to, Uint8 blend);
void        KXL_ReadBitmapHeader(const char *filename, KXL_BitmapHeader *hed);
Uint16      KXL_ReadU16(FILE *fp);
Uint32      KXL_ReadU32(FILE *fp);
#define     KXL_ReadS16 (Sint16)KXL_ReadU16
#define     KXL_ReadS32 (Sint32)KXL_ReadU16

//================================================================
// Global function prototypes
//================================================================
//================================================================
// Visual function prototypes
//================================================================
// Window functions
void        KXL_DisplayName(const char *name);
void        KXL_CreateWindow(Uint16 w, Uint16 h, const char *title, Uint32 event);
void        KXL_DeleteWindow(void);
// Frame functions
void        KXL_ReSizeFrame(Uint16 w, Uint16 h);
#define     KXL_UpDateRect(rr) KXL_Up_Date(rr.Left, rr.Top, rr.Width, rr.Height)
#define     KXL_UpDateImm KXL_Up_Date
#define     KXL_UpDate KXL_UpDateRect // Downward compatibility
#define     KXL_ClearFrameRect(rr) KXL_Clear_Frame(rr.Left, rr.Top, rr.Width, rr.Height)
#define     KXL_ClearFrameImm KXL_Clear_Frame
#define     KXL_ClearFrame KXL_ClearFrameRect // Downward compatibility
// Text functions
void        KXL_Font(const char *str, Uint8 r, Uint8 g, Uint8 b);
void        KXL_PutText(Sint16 x, Sint16 y, const char *str);
Uint16      KXL_TextWidth(const char *str);
// Drawing functions
void        KXL_SetDrawColor(Uint8 r, Uint8 g, Uint8 b);
void        KXL_DrawPoint(Sint16 left, Sint16 top);
void        KXL_DrawLine(Sint16 left, Sint16 top, Sint16 right, Sint16 bottom);
void        KXL_DrawRectangle(Sint16 left, Sint16 top, Uint16 width, Uint16 height, Bool flag);
void        KXL_DrawPolygon(KXL_Polygon *data, Uint16 max, Bool next, Bool flag);
// Image functions
KXL_Image  *KXL_LoadBitmap(const char *filename, Uint8 blend);
#define     KXL_CopyImageRect(src, rr) KXL_Copy_Image(src, rr.Left, rr.Top, rr.Width, rr.Height)
#define     KXL_CopyImageImm KXL_Copy_Image
#define     KXL_CopyImage KXL_CopyImageRect // Downward compatibility
#define     KXL_CopyStretchImageRect(src, rr, w, h) KXL_Copy_StretchImage(src, rr.Left, rr.Top, rr.Width, rr.Height, w, h)
#define     KXL_CopyStretchImageImm KXL_Copy_StretchImage
#define     KXL_CopyStretchImage(src, w, h) KXL_Copy_StretchImage(src, 0, 0, src->Width, src->Height, w, h)
#define     KXL_StrechImage KXL_CopyStretchImage // Downward compatibility
void        KXL_PutImage(KXL_Image *img, Sint16 x, Sint16 y);
#define     KXL_PutImageRect(src, rr, x, y) KXL_Put_Image(src, rr.Left, rr.Top, rr.Width, rr.Height, x, y)
#define     KXL_PutImageImm KXL_Put_Image
#define     KXL_PutRectImage KXL_PutImageRect // Downward compatibility
#define     KXL_PutStretchImage(src, w, h, x, y) KXL_Put_StretchImage(src, 0, 0, src->Width, src->Height, w, h, x, y)
#define     KXL_PutStretchImageRect(src, rr, w, h, x, y) KXL_Put_StretchImage(src, rr.Left, rr.Top, rr.Width, rr.Height, w, h, x, y)
#define     KXL_PutStretchImageImm KXL_Put_StretchImage
void        KXL_DeleteImage(KXL_Image *img);
// Event functions
Bool        KXL_CheckEvents(void);
Uint32      KXL_GetEvents(void);
Uint16      KXL_GetKey(void);
Uint16      KXL_GetButton(Uint16 *x, Uint16 *y);
//================================================================
// Sound function prototypes
//================================================================
void        KXL_InitSound(const char *path, char **fname);
void        KXL_EndSound(void);
void        KXL_PlaySound(Uint16 no, KXL_Command action);
//================================================================
// Etc. function prototypes
//================================================================
// Timer functions
void        KXL_Timer(Uint16 time);
void        KXL_ResetTimer(void);
Bool        KXL_GetTimer(void);
// Memory functions
void       *KXL_Malloc(Uint32 size);
void       *KXL_Realloc(void *src, Uint32 size);
void        KXL_Free(void *src);
// Game functions
Uint16      KXL_GetDirection(KXL_Rect src, KXL_Rect target);
void        KXL_GetDirectionAdd(Sint16 , Sint16 *x, Sint16 *y);
#define     KXL_DirectionAdd(dest, cnt, add) \
  cnt += add;\
  dest = cnt >> 8
#define     KXL_RectIntersect(m, y)\
  ((m.Left < y.Left + y.Width  - 1 &&\
    m.Top  < y.Top  + y.Height - 1 &&\
    y.Left < m.Left + m.Width  - 1 &&\
    y.Top  < m.Top  + m.Height - 1)\
    ? True : False)


#endif
